addons/mass_mailing/models/mailing_list.py

```diff
@@ -222,41 +246,73 @@ def action_merge(self, src_lists, archive):
+            FROM (
+                SELECT contact.id AS contact_id,
+                       list.id AS list_id,
+                       {select_values},
+                       {rn}
+                FROM mailing_contact contact,
+                     mailing_contact_list_rel contact_list_rel,
+                     mailing_list list
                 WHERE contact.id=contact_list_rel.contact_id
-                AND COALESCE(contact_list_rel.opt_out,FALSE) = FALSE
-                AND contact.email_normalized NOT IN (select email from mail_blacklist where active = TRUE)
+                AND {contact_condition}
                 AND list.id=contact_list_rel.list_id
-                AND list.id IN %s
-                AND NOT EXISTS
-                    (
-                    SELECT 1
-                    FROM
-                        mailing_contact contact2,
-                        mailing_contact_list_rel contact_list_rel2
-                    WHERE contact2.email = contact.email
-                    AND contact_list_rel2.contact_id = contact2.id
-                    AND contact_list_rel2.list_id = %s
+                AND list.id IN %(ids)s
+                AND ({select}) NOT IN (
+                    SELECT {join_select}
+                    FROM mailing_contact_list_rel rel
+                    JOIN mailing_contact sub_contact ON rel.contact_id=sub_contact.id
+                    WHERE rel.list_id = %(id)s
+                )
+            ) st
+            WHERE {rn_where};
+        """
+        update_query = f"""
+            UPDATE mailing_contact_list_rel
+            SET opt_out = TRUE
+            WHERE list_id = %(id)s
+            AND contact_id in (
+                SELECT id
+                FROM mailing_contact contact
+                WHERE ({select}) IN (
+                    SELECT {select.replace("contact.", "")}
+                    FROM mailing_contact
+                    WHERE id IN (
+                        SELECT contact_id
+                        FROM mailing_contact_list_rel
+                        WHERE COALESCE(mailing_contact_list_rel.opt_out, FALSE) = TRUE
+                        AND list_id IN %(ids)s
                     )
-                ) st
-            WHERE st.rn = 1;""", (self.id, tuple(src_lists.ids), self.id))
-        self.env.invalidate_all()
-        if archive:
-            (src_lists - self).action_archive()
+                )
+            );
+        """
+        self.env.cr.execute(insert_query, ({'id': self.id, 'ids': tuple(src_lists.ids)}))
+        self.env.cr.execute(update_query, ({'id': self.id, 'ids': tuple(src_lists.ids)}))
+        self.flush_model()
+        self.invalidate_model()
+        (src_lists - self).unlink()
+
+    def _mailing_list_get_contact_condition(self):
+        return "contact.email_normalized NOT IN (select email from mail_blacklist where active = TRUE)"
+
+    def _mailing_list_get_select_values(self):
+        return "contact.email"
```

Comment:
this should just return `['email']`, `contact` is the alias of the table in the SQL query, and should be added when we build the query (we already need to do it for `sub_contact`)

and it will also remove the `select.replace("contact.", "")` thing